### ISSUE TITLE
no-return-assign "always" still allows assignment if using OR in return statement

### ISSUE THREAD

**Issue by jedateach** [eslint/eslint#5913]
**What version of ESLint are you using?**
eslint@1.10.3 AND eslint@2.8.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```
{
	// ESLint Configuration File
	// see eslint.org for more details

	// 0 = off
	// 1 = warning - gradually introduced
	// 2 = error - breaks build

	// ensure this is the top-most eslint config file
	"root": true,

	// Rules as per http://eslint.org/docs/rules
	"rules": {

	// Possible Errors
	// The following rules point out areas where you might have made mistakes.

		// disallow or enforce trailing commas (recommended)
		"comma-dangle" : 2,
		// disallow assignment in conditional expressions (recommended)
		"no-cond-assign": [2, "always"],
		// disallow use of console in the node environment (recommended)
		"no-console": 2,
		// disallow use of constant expressions in conditions (recommended)
		"no-constant-condition": 2,
		// disallow control characters in regular expressions (recommended)
		"no-control-regex": 2,
		// disallow use of debugger (recommended)
		"no-debugger": 2,
		// disallow duplicate arguments in functions (recommended)
		"no-dupe-args": 2,
		// disallow duplicate keys when creating object literals (recommended)
		"no-dupe-keys": 2,
		// disallow a duplicate case label. (recommended)
		"no-duplicate-case": 2,
		// disallow the use of empty character classes in regular expressions (recommended)
		"no-empty-character-class": 2,
		// disallow empty statements (recommended)
		"no-empty": 2,
		// disallow assigning to the exception in a catch block (recommended)
		"no-ex-assign": 2,
		// disallow double-negation boolean casts in a boolean context (recommended)
		"no-extra-boolean-cast": 2,
		// disallow unnecessary parentheses
		"no-extra-parens": [2, "functions"],
		// disallow unnecessary semicolons (recommended) (fixable)
		"no-extra-semi": 2,
		// disallow overwriting functions written as function declarations (recommended)
		"no-func-assign": 2,
		// disallow function or variable declarations in nested blocks (recommended)
		"no-inner-declarations": 2,
		// disallow invalid regular expression strings in the RegExp constructor (recommended)
		"no-invalid-regexp": 2,
		// disallow irregular whitespace outside of strings and comments (recommended)
		"no-irregular-whitespace": 2,
		// disallow negation of the left operand of an in expression (recommended)
		"no-negated-in-lhs": 2,
		// disallow the use of object properties of the global object (Math and JSON) as functions (recommended)
		"no-obj-calls": 2,
		// disallow multiple spaces in a regular expression literal (recommended)
		"no-regex-spaces": 2,
		// disallow sparse arrays (recommended)
		"no-sparse-arrays": 2,
		// Avoid code that looks like two expressions but is actually one
		"no-unexpected-multiline": 2,
		// disallow unreachable statements after a return, throw, continue, or break statement (recommended)
		"no-unreachable": 2,
		// disallow comparisons with the value NaN (recommended)
		"use-isnan": 2,
		// Ensure JSDoc comments are valid
		"valid-jsdoc": [1, {
			"prefer": {
				"return": "returns"
			},
			"requireReturn": false,
			"requireParamDescription": false,
			"requireReturnDescription": false,
			"requireReturnType": true
		}],
		// Ensure that the results of typeof are compared against a valid string (recommended)
		"valid-typeof": 2,

	// Best Practices
	// These are rules designed to prevent you from making mistakes. They either prescribe a better way of doing something or help you avoid footguns.

		// Enforces getter/setter pairs in objects
		"accessor-pairs": 2,
		// treat var statements as if they were block scoped
		"block-scoped-var": 2,
		// specify the maximum cyclomatic complexity allowed in a program
		"complexity": [1, 10],
		// require return statements to either always or never specify values
		"consistent-return": 2,
		// specify curly brace conventions for all control statements
		"curly": 2,
		// require default case in switch statements
		"default-case": 2,
		// enforces consistent newlines before or after dots
		"dot-location": [2, "property"],
		// encourages use of dot notation whenever possible
		"dot-notation": [2, {"allowKeywords": false}],
		// require the use of === and !== (fixable)
		"eqeqeq": 2,
		// make sure for-in loops have an if statement
		"guard-for-in": 2,
		// disallow the use of alert, confirm, and prompt
		"no-alert": 2,
		// disallow use of arguments.caller or arguments.callee
		"no-caller": 2,
		// disallow lexical declarations in case clauses
		"no-case-declarations": 2,
		// disallow division operators explicitly at beginning of regular expression
		"no-div-regex": 2,
		// disallow else after a return in an if
		"no-else-return": 2,
		// disallow use of labels for anything other than loops and switches
		"no-empty-pattern": 2,
		// disallow comparisons to null without a type-checking operator
		"no-eq-null": 2,
		// disallow use of eval()
		"no-eval": 2,
		// disallow adding to native types
		"no-extend-native": 2,
		// disallow unnecessary function binding
		"no-extra-bind": 2,
		// disallow fallthrough of case statements (recommended)
		"no-fallthrough": 2,
		// disallow the use of leading or trailing decimal points in numeric literals
		"no-floating-decimal": 2,
		// disallow the type conversions with shorter notations
		"no-implicit-coercion": 2,
		// disallow use of eval()-like methods
		"no-implied-eval": 2,
		// disallow this keywords outside of classes or class-like objects
		"no-invalid-this": 2,
		// disallow usage of __iterator__ property
		"no-iterator": 2,
		// disallow use of labeled statements
		"no-labels": 2,
		// disallow unnecessary nested blocks
		"no-lone-blocks": 2,
		// disallow creation of functions within loops
		"no-loop-func": 2,
		// disallow the use of magic numbers
		"no-magic-numbers": 0,
		// disallow use of multiple spaces (fixable)
		"no-multi-spaces": 2,
		// disallow use of multiline strings
		"no-multi-str": 2,
		// disallow reassignments of native objects
		"no-native-reassign": 2,
		// disallow use of new operator for Function object
		"no-new-func": 2,
		// disallows creating new instances of String,Number, and Boolean
		"no-new-wrappers": 2,
		// disallow use of the new operator when not part of an assignment or comparison
		"no-new":  2,
		// disallow use of octal escape sequences in string literals, such as var foo = "Copyright \251";
		"no-octal-escape": 2,
		// disallow use of octal literals (recommended)
		"no-octal": 2,
		// disallow reassignment of function parameters
		// "no-param-reassign": 0,
		// disallow use of process.env
		"no-process-env": 2,
		// disallow usage of __proto__ property
		"no-proto": 2,
		// disallow declaring the same variable more than once (recommended)
		"no-redeclare": 2,
		// disallow use of assignment in return statement
		"no-return-assign": [2, "always"],
		// disallow use of javascript: urls.
		"no-script-url": 2,
		// disallow comparisons where both sides are exactly the same
		"no-self-compare": 2,
		// disallow use of the comma operator
		"no-sequences": 2,
		// restrict what can be thrown as an exception
		"no-throw-literal": 2,
		// disallow usage of expressions in statement position
		"no-unused-expressions": [2, {"allowShortCircuit": true, "allowTernary": true}],
		// disallow unnecessary .call() and .apply()
		"no-useless-call": 2,
		// disallow unnecessary concatenation of literals or template literals
		"no-useless-concat": 2,
		// disallow use of the void operator
		"no-void": 2,
		// disallow usage of configurable warning terms in comments - e.g. TODO or FIXME
		"no-warning-comments": 1,
		// disallow use of the with statement
		"no-with": 2,
		// require use of the second argument for parseInt()
		"radix": 2,
		// require declaration of all vars at the top of their containing scope
		"vars-on-top": 0,
		// require immediate function invocation to be wrapped in parentheses
		"wrap-iife": 2,
		// require or disallow Yoda conditions
		"yoda": 2,

	// Strict Mode
	// These rules relate to using strict mode.

		// controls location of Use Strict Directives
		// "strict": 0,

	// Variables
	// These rules have to do with variable declarations.

		// enforce or disallow variable initializations at definition
		// "init-declarations": 0,
		// disallow the catch clause parameter name being the same as a variable in the outer scope
		// "no-catch-shadow": 0,
		// disallow deletion of variables (recommended)
		// "no-delete-var": 0,
		// disallow labels that share a name with a variable
		// "no-label-var": 0,
		// disallow shadowing of names such as arguments
		// "no-shadow-restricted-names": 0,
		// disallow declaration of variables already declared in the outer scope
//		"no-shadow":  2,
		// disallow use of undefined when initializing variables
		// "no-undef-init": 0,
		// disallow use of undeclared variables unless mentioned in a /*global */ block (recommended)
//		"no-undef":  2,
		// disallow use of undefined variable
//		"no-undefined": 2,
		// disallow declaration of variables that are not used in the code (recommended)
//		"no-unused-vars":  2,
		// disallow use of variables before they are defined
		// "no-use-before-define": 0,

	// Node.js and CommonJS
	// These rules are specific to JavaScript running on Node.js or using CommonJS in the browser.

		// enforce return after a callback
		// "callback-return": 0,
		// enforce require() on top-level module scope
		// "global-require": 0,
		// enforce error handling in callbacks
		// "handle-callback-err": 0,
		// disallow mixing regular variable and require declarations
		// "no-mixed-requires": 0,
		// disallow use of new operator with the require function
		// "no-new-require": 0,
		// disallow string concatenation with __dirname and __filename
		// "no-path-concat": 0,
		// disallow process.exit()
		// "no-process-exit": 0,
		// restrict usage of specified node modules
		// "no-restricted-modules": 0,
		// disallow use of synchronous methods
		// "no-sync": 0,

	// Stylistic Issues
	// These rules are purely matters of style and are quite subjective.

		// enforce spacing inside array brackets (fixable)
		// "array-bracket-spacing": 0,
		// disallow or enforce spaces inside of single line blocks (fixable)
		// "block-spacing": 0,
		// enforce one true brace style
		// "brace-style": 0,
		// require camel case names
		// "camelcase": 0,
		// enforce spacing before and after comma (fixable)
		// "comma-spacing": 0,
		// enforce one true comma style
//		"comma-style": [2,"last"],
		// require or disallow padding inside computed properties (fixable)
		// "computed-property-spacing": 0,
		// enforce consistent naming when capturing the current execution context
		// "consistent-this": 0,
		// enforce newline at the end of file, with no multiple empty lines (fixable)
		// "eol-last": 0,
		// require function expressions to have a name
		// "func-names": 0,
		// enforce use of function declarations or expressions
		// "func-style": 0,
		// this option enforces minimum and maximum identifier lengths (variable names, property names etc.)
		// "id-length": 0,
		// require identifiers to match the provided regular expression
		// "id-match": 0,
		// specify tab or space width for your code (fixable)
//		"indent": [1,"tab",{"SwitchCase": 1}],
		// specify whether double or single quotes should be used in JSX attributes
		// "jsx-quotes": 0,
		// enforce spacing between keys and values in object literal properties
		// "key-spacing": 0,
		// disallow mixed 'LF' and 'CRLF' as linebreaks
		// "linebreak-style": 0,
		// enforce empty lines around comments
//		"lines-around-comment": [2, { "beforeBlockComment": true, "beforeLineComment": true }],
		// specify the maximum depth callbacks can be nested
		// "max-nested-callbacks": 0,
		// require a capital letter for constructors
		// "new-cap": 0,
		// disallow the omission of parentheses when invoking a constructor with no arguments
		// "new-parens": 0,
		// require or disallow an empty newline after variable declarations
		// "newline-after-var": 0,
		// disallow use of the Array constructor
		// "no-array-constructor": 0,
		// disallow use of the continue statement
		// "no-continue": 0,
		// disallow comments inline after code
		// "no-inline-comments": 0,
		// disallow if as the only statement in an else block
		// "no-lonely-if": 0,
		// disallow mixed spaces and tabs for indentation (recommended)
		// "no-mixed-spaces-and-tabs": 0,
		// disallow multiple empty lines
		// "no-multiple-empty-lines": 0,
		// disallow negated conditions
		// "no-negated-condition": 0,
		// disallow nested ternary expressions
		// "no-nested-ternary": 0,
		// disallow the use of the Object constructor
		// "no-new-object": 0,
		// disallow use of certain syntax in code
		// "no-restricted-syntax": 0,
		// disallow space between function identifier and application (fixable)
		// "no-spaced-func": 0,
		// disallow the use of ternary operators
		// "no-ternary": 0,
		// disallow trailing whitespace at the end of lines (fixable)
		// "no-trailing-spaces": 0,
		// disallow dangling underscores in identifiers
		// "no-underscore-dangle": 0,
		// disallow the use of ternary operators when a simpler alternative exists
		// "no-unneeded-ternary": 0,
		// require or disallow padding inside curly braces (fixable)
		// "object-curly-spacing": 0,
		// require or disallow one variable declaration per function
//		"one-var": [1, "never"],
		// require assignment operator shorthand where possible or prohibit it entirely
		// "operator-assignment": 0,
		// enforce operators to be placed before or after line breaks
		// "operator-linebreak": 0,
		// enforce padding within blocks
		// "padded-blocks": 0,
		// require quotes around object literal property names
		"quote-props": [2, "as-needed", { "keywords": true, "unnecessary": false }]
		// specify whether backticks, double or single quotes should be used (fixable)
//		"quotes":  [2,"double"],
		// Require JSDoc comment
		// "require-jsdoc": 0,
		// enforce spacing before and after semicolons
		// "semi-spacing": 0,
		// require or disallow use of semicolons instead of ASI (fixable)
//		"semi":  1,
		// sort variables within the same declaration block
		// "sort-vars": 0,
		// require a space after certain keywords (fixable)
		// "space-after-keywords": 0,
		// require or disallow a space before blocks (fixable)
//		"space-before-blocks": 1,
		// require or disallow a space before function opening parenthesis (fixable)
		// "space-before-function-paren": 0,
		// require a space before certain keywords (fixable)
		// "space-before-keywords": 0,
		// require or disallow spaces inside parentheses
		// "space-in-parens": 0,
		// require spaces around operators (fixable)
//		"space-infix-ops":  2,
		// require a space after return, throw, and case (fixable)
		// "space-return-throw-case": 0,
		// require or disallow spaces before/after unary operators (fixable)
		// "space-unary-ops": 0,
		// require or disallow a space immediately following the // or /* in a comment
//		"spaced-comment":  1,
		// require regex literals to be wrapped in parentheses
		// "wrap-regex": 0,

	// ECMAScript 6
	// These rules are only relevant to ES6 environments.

		// require braces in arrow function body
		// "arrow-body-style": 0,
		// require parens in arrow function arguments
		// "arrow-parens": 0,
		// require space before/after arrow function's arrow (fixable)
		// "arrow-spacing": 0,
		// verify calls of super() in constructors
		// "constructor-super": 0,
		// enforce spacing around the * in generator functions (fixable)
		// "generator-star-spacing": 0,
		// disallow arrow functions where a condition is expected
		// "no-arrow-condition": 0,
		// disallow modifying variables of class declarations
		// "no-class-assign": 0,
		// disallow modifying variables that are declared using const
		// "no-const-assign": 0,
		// disallow duplicate name in class members
		// "no-dupe-class-members": 0,
		// disallow use of this/super before calling super() in constructors.
		// "no-this-before-super": 0,
		// require let or const instead of var
		// "no-var": 0,
		// require method and property shorthand syntax for object literals
		// "object-shorthand": 0,
		// suggest using arrow functions as callbacks
		// "prefer-arrow-callback": 0,
		// suggest using const declaration for variables that are never modified after declared
		// "prefer-const": 0,
		// suggest using Reflect methods where applicable
		// "prefer-reflect": 0,
		// suggest using the spread operator instead of .apply().
		// "prefer-spread": 0,
		// suggest using template literals instead of strings concatenation
		// "prefer-template": 0,
		// disallow generator functions that do not have yield
		// "require-yield": 0,

	// Legacy
	// The following rules are included for compatibility with JSHint and JSLint. While the names of the rules may not match up with the JSHint/JSLint counterpart, the functionality is the same.

		// specify the maximum depth that blocks can be nested
		// "max-depth": 0,
		// specify the maximum length of a line in your program
		// "max-len": 0,
		// limits the number of parameters that can be used in the function declaration.
		// "max-params": 0,
		// specify the maximum number of statement allowed in a function
		// "max-statements": 0,
		// disallow use of bitwise operators
//		"no-bitwise": 2,
		// disallow use of unary operators, ++ and --
		// "no-plusplus":  0,

	// Removed
	// These rules existed in a previous version of ESLint but have since been replaced by newer rules.

		// enforce the position of the * in generator functions (replaced by generator-star-spacing)
		// "generator-star": 0,
		// require or disallow the "use strict" pragma in the global scope (replaced by strict)
		// "global-strict": 0,
		// disallow trailing commas in object literals (replaced by comma-dangle)
		// "no-comma-dangle": 0,
		// disallow the use of empty character classes in regular expressions (replaced by no-empty-character-class)
		// "no-empty-class": 0,
		// disallow unnecessary use of "use strict"; when already in strict mode (replaced by strict)
		// "no-extra-strict": 0,
		// disallow reserved words being used as object literal keys
		// "no-reserved-keys": 0,
		// disallow space before semicolon (replaced by semi-spacing)
		// "no-space-before-semi": 0,
		// disallow wrapping of non-IIFE statements in parens (replaced by no-extra-parens)
		// "no-wrap-func": 0,
		// require a space after function names (replaced by space-before-function-paren)
		// "space-after-function-name": 0,
		// require or disallow space before function parentheses (replaced by space-before-function-paren)
		// "space-before-function-parentheses": 0,
		// require or disallow spaces inside brackets (replaced by object-curly-spacing and array-bracket-spacing)
		// "space-in-brackets": 0,
		// require or disallow spaces before/after unary operators (replaced by space-unary-ops)
		// "space-unary-word-ops": 0,
		// require or disallow a space immediately following the // in a line comment (replaced by spaced-comment)
		// "spaced-line-comment": 0,
	},

	"parserOptions": {
        "ecmaVersion": 3
    },

	// Environments
	"env": {
		"browser": true,			// Web Browser (window, document, etc)
		"jquery": true				// jQuery
	},

	// Custom Globals
	"globals": {
		"devModeOnly": true,
		"Modernizr": true,
		"require": true,
		"requirejs": true,
		"define": true,
		"IBPage": true,
		"Ember": true
	}
}

```

**What did you do? Please include the actual source code causing the issue.**
```
var fn = (function() {
	var a = true;
	if (a) {
		return (a = "thing"); // has eslint 'no-return-assign' error
	}
	if (a === 3) {
		return a || (a = "thing"); // no 'no-return-assign' error here
	}
	return true;
}());

```

**What did you expect to happen?**
A 'no-return-assign' error

**What actually happened? Please include the actual, raw output from ESLint.**
No error found.

```
$ grunt eslint:errors
Running "eslint:errors" (eslint) task

/myproject/eslinttestfile.js
  4:3  error  Return statement should not contain assignment  no-return-assign

✖ 1 problem (1 error, 0 warnings)

Warning: Task "eslint:errors" failed. Use --force to continue.
Error: Task "eslint:errors" failed.
    at Task.<anonymous> (/myproject/node_modules/grunt/lib/util/task.js:205:15)
    at Task.runTaskFn (/myproject/node_modules/grunt/lib/util/task.js:254:9)
    at Task.<anonymous> (/myproject/node_modules/grunt/lib/util/task.js:300:12)
    at Task.start (/myproject/node_modules/grunt/lib/util/task.js:309:5)
    at Object.grunt.tasks (/myproject/node_modules/grunt/lib/grunt.js:164:8)
    at Object.module.exports [as cli] (/myproject/node_modules/grunt/lib/grunt/cli.js:38:9)
    at Object.<anonymous> (/usr/local/lib/node_modules/grunt-cli/bin/grunt:45:20)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
    at node.js:814:3

Aborted due to warnings.
```
...note no error present for line 7.

**Comment by michaelficarra** [comment]
I don't think this rule is meant to be that smart. Would you also expect it to catch this?

```js
(function(){
  var a = 0;
  return (function(){ a = 1; return a; }());
}());
```

**Comment by jedateach** [comment]
I don't think so, as that looks like a legitimate assignment before returning a value...rather than an assignment as part of an expression.

**Comment by michaelficarra** [comment]
But there's an assignment within the returned expression. So do you stop traversing at function boundaries?

**Comment by jedateach** [comment]
Yeah, that is my assumption.
Generalising a bit: I thought the rule would prevent assignments in situations where they can be confused with comparisons.

**Comment by mysticatea** [comment]
`no-cond-assign` rule is warning both patterns. So I think we should fix `no-return-assign` rule.

![image](https://cloud.githubusercontent.com/assets/1937871/14872723/1543638e-0d2a-11e6-8452-80f94bfbe8bc.png)

**Comment by mysticatea** [comment]
I'm working on this.
